src/Extension/ExtensionManager.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+namespace Carnival\Extension;
+
+use Carnival\Extension\Contracts\ExtensionManager as ExtensionManagerContract;
+use Carnival\Extension\Contracts\Extension;
+use Carnival\Extension\Contracts\Installable;
+use Illuminate\Support\Collection;
+use InvalidArgumentException;
+
+class ExtensionManager implements ExtensionManagerContract {
+    private Collection $extensions;
+
+    public function __construct() {
+        $this->extensions = new Collection();
+    }
+
+    public function register(Extension $extension) : void {
+        $this->extensions->put(get_class($extension), $extension);
+    }
+
+    public function isRegistered(Extension $extension) : bool {
+        return $this->extensions->has(get_class($extension));
+    }
+
+    public function getExtensions() : Collection {
+        return $this->extensions->values();
+    }
+
+    public function activate(Extension $extension) : void {
+        $this->actOnRegisteredExtension($extension, fn(Extension $extension) => $extension->activate());
+    }
+
+    public function deactivate(Extension $extension) : void {
+        $this->actOnRegisteredExtension($extension, fn(Extension $extension) => $extension->deactivate());
+    }
+
+    public function install(Extension $extension) : void {
+        $this->actIfInstallable($extension, fn(Extension&Installable $extension) => $extension->install());
```

Comment:
I remember we discussed this syntax before but I already forget. Is it that it can be either type? Maybe leave a comment since it is a rare occurance

Reply:
> I remember we discussed this syntax before but I already forget. Is it that it can be either type?

It's an & (and), think of it like an intersection. `$extension` must be both

> Maybe leave a comment since it is a rare occurance

I lean against putting comments in source code unless it's really necessary. This is a documented PHP feature after all
Plus, just from reading the code, it should be deducible. Per line 38, it's an Extension. Then, once passing through `actIfInstallable`, now it's an "Extension&Installable" or, in English, an "Extension AND Installable"

Reply:
Ah, okay. Sounds good then if you don't want to comment it then.